adding typespecs

gausby opened the pull request:
I have added some typespecs.

They are not complete yet, I miss `handle_event/3`, `handle_sync_event/4` `handle_info/3`, and `code_change/4`.

I have looked at the implementation for the Elixir core gen_server wrapper, and I've tried to follow the Erlang documentation on the different specs for the functions. I am not 100% sure they are correct.

It's a start. You can let this pull request linger and I'll get back to adding specs for the remaining callbacks. Perhaps I need to reorder some of the specs and types, and use the types for state_name and state_data consistently.

gausby committed:
Adding type-specs and callbacks

- Minor clean up, removed an seemingly unused `@doc false`. I don't
  think it applied to anything. Fixed indentation of a line, etc

- Changed the default implementations to versions that will kill the
  server if unknown events are received. This should be more in line
  with the standard behavior of other Elixir gen-wrappers (sending a
  gen_server a message it does not know how to handle kill the server)

- Changed the unit tests to handle {:EXIT, *pid*, :unexpected_event},
  the other data should be written to the error log anyways

- Added the remaining callbacks

- The type definitions could perhaps be moved around, feel free to do so
  if one can make it look prettier

- Added a turn-stile example that also serves as a test for sync events

- Please have a look at the specs and add to them, or fix them if they
  are incorrect or not precise

pavlos commented:
wow, this is great.  Thank you so much!  I will let the PR sit until you want me to merge it

gausby commented:
Thanks. Let me know if you see something that is wrong :)

gausby commented:
I've added the remaining callbacks, and I've added the turnstile example as a test because it test the sync calls—the commits has been squashed into one, so it shouldn't mess up the history too much.

Merge it if you find it good.

pavlos commented on lib/gen_fsm.ex:
```diff
@@ -20,69 +85,82 @@ defmodule GenFSM do
     end
   end
 
+  @spec stop(server, reason :: term, timeout) :: :ok
```
should this be fsm_ref instead of server?

pavlos commented on lib/gen_fsm.ex:
```diff
@@ -20,69 +85,82 @@ defmodule GenFSM do
     :gen.stop(fsm, reason, timeout)
   end
 
+  @type fsm_ref :: pid | name
```
and then @type fsm_ref should be the same as @server above, and we can just eliminate @server?

gausby replied:
I am actually kind confused about that

Server is defined as `@type server :: pid | name | {atom, node}` and fsm_ref is defined to be `@type fsm_ref :: pid | name`. I don't know if it allows for `{atom, node}` the places it has been set to be an `fsm_ref`.

pavlos replied:
I think it will allow for it - I checked the erlang docs and http://erlang.org/doc/man/gen_fsm.html#send_event-2 and http://erlang.org/doc/man/gen_fsm.html#stop-3 have FsmRef listed and allow for  Name | {Name,Node} | {global,GlobalName} | {via,Module,ViaName} | pid().  

What do you think?

pavlos commented on lib/gen_fsm.ex:
```diff
@@ -20,69 +85,82 @@ defmodule GenFSM do
       @doc false
-      def handle_info(_msg, state_name, state_data) do
-        { :next_state, state_name, state_data }
+      def handle_info(_info, _state_name, state_data) do
```
i like this - better to crash on an unexpected message :+1:

pavlos commented on test/turnstile_example_test.exs:
```diff
@@ -0,0 +1,92 @@
+defmodule GenFSM.TurnstileExampleTest do
```
this is great, thank you!

gausby replied:
We could move this into a different module. This naming was just the first that came to mind.

We also need to do more unit tests; the turnstile is more of an integration test.

pavlos commented on lib/gen_fsm.ex:
```diff
@@ -20,69 +85,82 @@ defmodule GenFSM do
-      def handle_event(event, state_name, state_data) do
-        { :stop, {:bad_event, state_name, event}, state_data }
+      def handle_event(_event, _state_name, state_data) do
+        {:stop, :unexpected_event, state_data}
```
wouldn't it be helpful to return the state_name and event in case of an unexpected event?  That would make it easier to debug?  Or is there a good reason you removed it?

gausby replied:
I think that the crash report contains all the information needed for debugging. I think it is redundant to pass it along, and if needed one could just define the needed callback—in fact, if one implement a callback one need to implement the fall through case anyways (as I understand)

pavlos commented on lib/gen_fsm.ex:
```diff
@@ -20,69 +85,82 @@ defmodule GenFSM do
-      def handle_sync_event(event, from, state_name, state_data) do
-        { :stop, {:bad_sync_event, state_name, event}, state_data }
+      def handle_sync_event(_event, _from, _state_name, state_data) do
+        {:stop, :unexpected_event, state_data}
```
see above

pavlos commented:
@gausby this is awesome, please check out the comments I left and let me know what you think.

gausby commented:
Thanks for the feedback. I've replied to most of them!

gausby commented:
I've fixed the `server` vs `fsm_ref`, I went with fsm_ref :)

I am going to squash the commit and force push it to the branch to keep the history clean.

gausby commented:
I guess the only outstanding issue is wether or not to include state name and event name with the default callback implementation of event and sync event callbacks. I would still argue that the crash report will contain this information, so we don't need to pass it to the return tuple. `{:stop, :unexpected_event, state_data}` will quite simply do in my opinion.

That is how I have my current gen_fsm yasnippet setup in my editor, as seen here: https://github.com/gausby/mg-elixir-snippets/blob/master/snippets/elixir-mode/behaviour-gen-fsm :)

pavlos commented:
let's leave out the state/event name, and we can always add it back in if it becomes a problem

pavlos commented:
merged!
